### ISSUE TITLE
How to prevent the host from entering TD through ssh?

### ISSUE THREAD

**Issue by zibinpan** [canonical/tdx#92]
Hello, thanks for your thoughtful work. I have a simple question:

Since TD is created on the host guest, set the `ssh` to enter without setting a password. So it seems that the host rooter (an untrusted guy) can use `ssh` to directly enter the TD to do some bad things, e.g., change some kernel files. So how can we still trust this TD? Although we can reset the password, something of the OS may have already been changed by the untrusted host rooter.

In another way, if we set the password when starting the TD, but during this time we still in an untrusted environment without protected by TDX, so the password may be peeped by the untrusted host rooter. Then after entering the TD, the host can also enter my TD, too.

In conclusion, how can we ensure or trust our TD is safe? Especially in preventing the host rooter from entering.

**Comment by matti** [comment]
> In conclusion, how can we ensure or trust our TD is safe? Especially in preventing the host rooter from entering.

So you'd run the attestation, but then again how to verify that --> his is essentially exactly what I'm asking in #98

**Comment by hector-cao** [comment]
@matti @zibinpan The goal of this repository is to allow you to enable TDX on Ubuntu for appropriate hardware by installing some specific software components (kernel, qemu, attestation, ...), it does not aim to provide you an end-2-end secure solution for any of the possible TDX use cases, the guest image creation is only here to help users to quickly create a guest to play with,

**Comment by BFuhry** [comment]
There are multiple options to prevent the infrastructure provider from logging into the Trust Domain (TD). Intel TDX does not stipulate any way of achieving this goal, but I will explain two possible options in the following. Note these options assume that you are in full control of the TD image, which might not be the case with some infrastructure providers.

One easy option would be to put SSH public keys into the TD image. As a result, only users with access to the corresponding private keys will have access to the TD and the public keys are not confidential. To be sure that nobody - including the infrastructure provider - changed the image (including they SSH pubic keys), remote attestation has to be performed. As the keys are part of the image, they will be part of the TD measurement and any change would be noticeable.

Another option is to deny all OS-based login capabilities (for anybody). Instead, put a software/agent inside the TD offering an endpoint handling all communication. This software/agent could enforce authentication (and authorization) by, e.g., only accepting client certificates signed by a hard-coded CA or by using some external, trusted authentication service.
